lms/static/scripts/frontend_apps/components/test/BookSelector-test.js

```diff
@@ -0,0 +1,224 @@
+import { mount } from 'enzyme';
+import { createElement } from 'preact';
+
+import { waitFor, waitForElement } from '../../../test-util/wait';
+import { VitalSourceService, withServices } from '../../services';
+import * as bookData from '../../utils/vitalsource-sample-data';
+import BookSelector, { $imports } from '../BookSelector';
+
+describe('BookSelector', () => {
+  let fakeVitalSourceService;
+  let fakeBookIDFromURL;
+
+  const BookSelectorWrapper = withServices(BookSelector, () => [
+    [VitalSourceService, fakeVitalSourceService],
+  ]);
+
+  const renderBookSelector = (props = {}) =>
+    mount(<BookSelectorWrapper onSelectBook={sinon.stub()} {...props} />);
+
+  beforeEach(() => {
+    fakeBookIDFromURL = sinon.stub().returns('BOOKSHELF_TUTORIAL');
+    fakeVitalSourceService = {
+      fetchBook: sinon.stub().callsFake(async () => bookData.bookList[0]),
+    };
+
+    $imports.$mock({
+      '../utils/vitalsource': {
+        bookIDFromURL: fakeBookIDFromURL,
+      },
+    });
+  });
+
+  afterEach(() => {
+    $imports.$restore();
+  });
+
+  const getInput = wrapper => wrapper.find('TextInput').find('input');
+
+  // Set the value of the input field but do not fire any events
+  const setURL = (wrapper, url) => {
+    const input = getInput(wrapper);
+    input.getDOMNode().value = url;
+  };
+
+  // Set the value of the input field AND fire `change` event
+  const updateURL = (wrapper, url) => {
+    setURL(wrapper, url);
+    const input = getInput(wrapper);
+    input.simulate('change');
+  };
+
+  context('entering, changing and submitting book URL', () => {
+    it('validates entered URL when the value of the text input changes', () => {
+      const wrapper = renderBookSelector();
+
+      updateURL(wrapper, 'foo');
+
+      assert.calledOnce(fakeBookIDFromURL);
+      assert.calledWith(fakeBookIDFromURL, 'foo');
+
+      updateURL(wrapper, 'bar');
+
+      assert.equal(
+        fakeBookIDFromURL.callCount,
+        2,
+        're-validates if entered URL value changes'
+      );
+      assert.calledWith(fakeBookIDFromURL, 'bar');
+
+      updateURL(wrapper, 'bar');
+
+      assert.equal(
+        fakeBookIDFromURL.callCount,
+        2,
+        'Does not validate URL if it has not changed from previous value'
+      );
+    });
+
+    it('validates entered URL when input is focused and "Enter" is pressed', () => {
+      const wrapper = renderBookSelector();
+      const input = getInput(wrapper);
+      const keyEvent = new Event('keydown');
+      keyEvent.key = 'Enter';
+
+      setURL(wrapper, 'http://www.example.com');
+
+      input.getDOMNode().dispatchEvent(keyEvent);
```

Comment:
As an alternative to creating a `keyEvent` manually, this also works:

```js
input.simulate('keydown', { key: 'Enter' })
```

Currently this will create a regular `Event` rather than a `KeyboardEvent`. Fortunately the code doesn't care about the distinction.